File: data-structures/6. doubly-linked-list/doubly-linked-list.js
//https://medium.com/better-programming/singly-and-doubly-linked-lists-in-javascript-7515f47c9f60s

//Doubly Linked List Node
class Node {
    constructor(value) {
        this.value = value;
        // at beginning, nothing comes after
        this.next = null;
        this.prev = null;
    }
}



class DoublyLinkedList {
    constructor() {
        this.head = null
        this.tail = null
        this.length = 0
    }

    isEmpty() {
        return this.length == 0
    }

    push(value) {
        const newNode = new Node(value)

        if (this.isEmpty()) {
            this.head = newNode
            this.tail = newNode
        } else {
            // our current tail is going to become
            // our 2nd to last node
            // it's next should point to our new tail -new Node
            this.tail.next = newNode
            newNode.prev = this.tail
            this.tail = newNode
        }
        this.length++
    }

    pop() {
        if (this.isEmpty()) {
            return null
        } else if (this.length == 1) {
            const nodeToRemove = this.head
            this.head = null
            this.tail = null
            return nodeToRemove
        } else {
            //there are mulitple nodes in the list
            let currentNode = this.head
            const nodeToRemove = this.tail
            //this will be our new tail
            let secondToLastNode
            while (currentNode) {
                if (currentNode.next == this.tail) {
                    secondToLastNode = currentNode
                    break
                }
                currentNode = currentNode.next
            }

            //our new tail will not have a next
            secondToLastNode.next = null
            // our current tail is being set to our new tail
            this.tail = secondToLastNode
            this.length--
            console.log('nodeToRemove', nodeToRemove)
            return nodeToRemove.value
        }
    }

    get(index) {
        if (index < 0 || index > this.length || this.isEmpty()) {
            return null
        }
        /* We want the first node */
        if (index == 0) {
            return this.head
        }

        /* We want the last node */
        if (index === this.length - 1) {
            return this.tail
        }

        /* We want a node somewhere in the list so we iterate*/

        let currentNode = this.head
        let iterator = 0
        while (iterator < index) {
            iterator++
            currentNode = currentNode.next
        }

        return currentNode
    }

    print() {
        let currNode = this.head
        while (currNode !== null) {
            console.log(currNode.value)
            currNode = currNode.next
        }
    }

}



const list = new DoublyLinkedList();
list.push("Emma");
list.push("Sarah");
list.push("Ivy");
list.print()
console.log(list.get(0).value) //Emma
console.log(list.pop()) //Ivy

File: data-structures/6. doubly-linked-list/doubly_linked_list.py
"""Each ListNode holds a reference to its previous node
as well as its next node in the List.



https://www.pythoncentral.io/reverse-singly-linked-list/
https://www.youtube.com/watch?v=MRe3UsRadKw

"""


class ListNode:
    def __init__(self, value, prev=None, next=None):
        self.value = value
        self.prev = prev
        self.next = next

"""Our doubly-linked list class. It holds references to
the list's head and tail nodes."""


class DoublyLinkedList:
    def __init__(self, node=None):
        self.head = node
        self.tail = node
        self.length = 1 if node is not None else 0

    def isEmpty(self):
        return self.length == 0

    def push(self, value):
        newNode = ListNode(value)

        if self.isEmpty():
            self.head = newNode
            self.tail = newNode
        else:
            #current tail will become 2nd to last node
            self.tail.next = newNode
            newNode.prev = self.tail
            self.tail = newNode
        self.length+=1
    

    def swap(self, node):
        prev = node.prev
        node.prev = node.next
        node.next = prev 

    def reverse(self):
        '''
        1. Swap prev and next pointers for all nodes
        2. Change prev of head and change head pointer in the e
        '''
        prev = None
        currNode = self.head
        while currNode:
            self.swap(currNode)
            #update prev node before moving to next
            prev = currNode
            #advance using prev pointer, since next & prev were swapped
            currNode = currNode.prev
  
        if prev:
            self.head = prev

        return self.head
    
  
    def print_it(self):
        currNode = self.head
        while currNode:
            print(currNode.value) 
            currNode = currNode.next


    ## we can also do the reverse with recursionl.,..
list = DoublyLinkedList()
list.push("Emma")
list.push("Sarah")
list.push("Ivy")
list.print_it()
# print(list.get(0).value) //Emma
# print(list.pop()) //Ivy
list.reverse()
list.print_it()
